File: net/tcp/py/tcp_server.py
#!/usr/bin/env python3
"""
TCP Server - Multi-threaded implementation
Accepts client connections and handles each in a separate thread
No external dependencies required
"""

import socket
import threading
import sys
from threading import Lock

# Configuration
HOST = '127.0.0.1'
PORT = 8888
BACKLOG = 5
BUFFER_SIZE = 1024

# Global variables for thread safety
client_counter = 0
counter_lock = Lock()

def handle_client(client_socket, client_id):
    """
    Handle a single client connection in a separate thread
    
    Args:
        client_socket: The socket connection to the client
        client_id: Unique identifier for this client
    """
    print(f"Client {client_id} connected. Socket: {client_socket.fileno()}")
    
    try:
        while True:
            # Receive data from client
            data = client_socket.recv(BUFFER_SIZE)
            
            if not data:
                print(f"Client {client_id} disconnected.")
                break
            
            message = data.decode('utf-8', errors='replace')
            print(f"Client {client_id} sent: {message}")
            
            # Send response back to client
            response = f"Server received: {message}"
            client_socket.sendall(response.encode('utf-8'))
            
    except Exception as e:
        print(f"Error handling client {client_id}: {e}")
    finally:
        client_socket.close()
        print(f"Client {client_id} connection closed.")


def start_server():
    """
    Start the TCP server and accept client connections
    Each connection is handled in a separate thread
    """
    global client_counter
    
    # Create server socket
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    # Allow address reuse
    server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    
    try:
        # Bind to address
        server_socket.bind((HOST, PORT))
        print(f"Server binding to {HOST}:{PORT}")
        
        # Listen for connections
        server_socket.listen(BACKLOG)
        print(f"Server listening on port {PORT}")
        print("Press Ctrl+C to stop the server\n")
        
        while True:
            try:
                # Accept client connection
                client_socket, client_address = server_socket.accept()
                
                with counter_lock:
                    client_counter += 1
                    current_client_id = client_counter
                
                print(f"New connection from {client_address[0]}:{client_address[1]}")
                
                # Create a new thread for this client
                client_thread = threading.Thread(
                    target=handle_client,
                    args=(client_socket, current_client_id),
                    daemon=False
                )
                client_thread.start()
                print(f"Started thread for Client {current_client_id}\n")
                
            except Exception as e:
                print(f"Error accepting connection: {e}")
                
    except KeyboardInterrupt:
        print("\n\nServer stopping...")
    except Exception as e:
        print(f"Server error: {e}")
    finally:
        server_socket.close()
        print("Server stopped.")


if __name__ == '__main__':
    try:
        start_server()
    except Exception as e:
        print(f"Fatal error: {e}")
        sys.exit(1)

File: net/tcp/cpp/tcp_server.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <string>
#include <cstring>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <atomic>
#include <memory>

const int PORT = 8888;
const int BACKLOG = 5;
const int BUFFER_SIZE = 1024;

std::atomic<bool> running(true);
std::vector<std::shared_ptr<std::thread>> threads;

// 处理客户端连接的线程函数
void handleClient(int clientSocket, int clientId) {
    std::cout << "Client " << clientId << " connected. Socket: " << clientSocket << std::endl;
    
    char buffer[BUFFER_SIZE];
    ssize_t bytesRead;
    
    try {
        while (running) {
            // 接收客户端数据
            bytesRead = recv(clientSocket, buffer, BUFFER_SIZE - 1, 0);
            
            if (bytesRead > 0) {
                buffer[bytesRead] = '\0';
                std::cout << "Client " << clientId << " sent: " << buffer << std::endl;
                
                // 回显数据给客户端
                std::string response = "Server received: " + std::string(buffer);
                send(clientSocket, response.c_str(), response.length(), 0);
                
            } else if (bytesRead == 0) {
                std::cout << "Client " << clientId << " disconnected." << std::endl;
                break;
            } else {
                std::cerr << "Error reading from client " << clientId << std::endl;
                break;
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception in handleClient: " << e.what() << std::endl;
    }
    
    // 关闭连接
    close(clientSocket);
    std::cout << "Client " << clientId << " connection closed." << std::endl;
}

// 主服务器函数
void startServer() {
    // 创建服务器套接字
    int serverSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (serverSocket < 0) {
        std::cerr << "Error creating server socket" << std::endl;
        return;
    }
    
    // 设置套接字选项，允许地址复用
    int opt = 1;
    if (setsockopt(serverSocket, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) < 0) {
        std::cerr << "Error setting socket options" << std::endl;
        close(serverSocket);
        return;
    }
    
    // 绑定地址
    struct sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = INADDR_ANY;
    serverAddr.sin_port = htons(PORT);
    
    if (bind(serverSocket, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
        std::cerr << "Error binding socket to port " << PORT << std::endl;
        close(serverSocket);
        return;
    }
    
    // 监听连接
    if (listen(serverSocket, BACKLOG) < 0) {
        std::cerr << "Error listening on socket" << std::endl;
        close(serverSocket);
        return;
    }
    
    std::cout << "Server listening on port " << PORT << std::endl;
    
    int clientId = 0;
    
    // 接受客户端连接
    while (running) {
        struct sockaddr_in clientAddr;
        socklen_t clientAddrLen = sizeof(clientAddr);
        
        int clientSocket = accept(serverSocket, (struct sockaddr*)&clientAddr, &clientAddrLen);
        
        if (clientSocket < 0) {
            if (running) {
                std::cerr << "Error accepting connection" << std::endl;
            }
            continue;
        }
        
        clientId++;
        
        // 为每个客户端创建新线程
        auto clientThread = std::make_shared<std::thread>(handleClient, clientSocket, clientId);
        threads.push_back(clientThread);
        
        std::cout << "Started thread for Client " << clientId << std::endl;
    }
    
    // 等待所有线程完成
    for (auto& t : threads) {
        if (t && t->joinable()) {
            t->join();
        }
    }
    
    close(serverSocket);
    std::cout << "Server stopped." << std::endl;
}

int main() {
    std::cout << "Starting TCP Server..." << std::endl;
    
    try {
        startServer();
    } catch (const std::exception& e) {
        std::cerr << "Exception in main: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}
